theano/compile/function_module.py

```diff
@@ -483,6 +488,66 @@ def __copy__(self):
                 there.data = here.data
         return cpy
 
+    def copy(self, share_memory = False):
+        """
+        Copy this function. Copied function will have separated maker and fgraph
+        with original function. User can choose whether to separate storage by 
+        changing the share_memory arguments
+        ---------------------
+        Params:
+            share_memory -- { boolean } Default is False. When True, two function
+            share intermediate storages(storages except input and output storages)
+        ---------------------
+        Returns:
+            func -- Copied theano.Function
+        """
+
+        if not share_memory:
+            return self.__copy__()
+        else:
+            # copy SymbolocKits
+            ins, outs = copy.deepcopy([self.maker.inputs, self.maker.outputs])
```

Comment:
```In.value```, ```storage_map[input_var]``` and storage in thunks shared the same reference of input storages. Therefore I think they should be copied. And the values should refer to same object(shared) if ```In.mutable == False and instance(In.variable, theano.tensor.Constant)```.

There was a bug in my code that I did not share the storage between new ```In``` and old ```In```. Now it's fixed.